sndio.c

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (C) 2011 Donovan "Tsomi" Watteau <tsoomi@gmail.com>
+ *
+ * Based on Thomas Pfaff's work for XMMS, and some suggestions from
+ * Alexandre Ratchov.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <sys/types.h>
+#include <sys/ioctl.h>
+#include <sys/audioio.h>
+#include <sys/stat.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+#include <unistd.h>
+#include <sndio.h>
+
+#include "op.h"
+#include "mixer.h"
+#include "sf.h"
+#include "xmalloc.h"
+
+static sample_format_t sndio_sf;
+static struct sio_par par;
+static struct sio_hdl *hdl = NULL;
+static int sndio_volume = 100;
+static int sndio_paused;
+
+static int sndio_mixer_set_volume(int l, int r)
+{
+	sndio_volume = l > r ? l : r;
+
+	if (hdl != NULL)
+		sio_setvol(hdl, sndio_volume * SIO_MAXVOL / 100);
+
+	return 0;
+}
+
+static int sndio_mixer_get_volume(int *l, int *r)
+{
+	*l = *r = sndio_volume;
+
+	return 0;
+}
+
+static int sndio_set_sf(sample_format_t sf)
+{
+	struct sio_par apar;
+
+	sndio_sf = sf;
+
+	sio_initpar(&par);
+
+	par.pchan = sf_get_channels(sndio_sf);
+	par.rate = sf_get_rate(sndio_sf);
+	sndio_paused = 0;
+
+	if (sf_get_signed(sndio_sf))
+		par.sig = 1;
+	else
+		par.sig = 0;
+
+	if (sf_get_bigendian(sndio_sf))
+		par.le = 0;
+	else
+		par.le = 1;
+
+	switch (sf_get_bits(sndio_sf)) {
+	case 16:
+		par.bits = 16;
+		break;
+	case 8:
+		par.bits = 8;
+		break;
+	default:
+		return -1;
+	}
+
+	par.appbufsz = par.rate * 300 / 1000;
+	apar = par;
+
+	if (!sio_setpar(hdl, &par))
+		return -1;
```

Comment:
Here and elsewhere: can you please return `-OP_ERROR_SOMETHING` instead of `-1`? It can be a nice idea to log some errors (with `d_print()`) too.